benches/dapol.rs

```diff
@@ -1,175 +1,250 @@
-                },
-                BatchSize::SmallInput,
-            );
-        });
+    thread_counts.push(max_thread_count);
+
+    let mut ndm_smt = Option::<NdmSmt>::None;
+
+    e.advance().unwrap();
+
+    for h in TREE_HEIGHTS.into_iter() {
+        e.advance().unwrap();
+
+        for t in thread_counts.iter() {
+            e.advance().unwrap();
+
+            for u in NUM_USERS.into_iter() {
+                e.advance().unwrap();
+
+                let max_users_for_height = 2_u64.pow((h - 1) as u32);
+
+                if u > max_users_for_height {
+                    break;
+                }
+
+                let tup: (Height, MaxThreadCount, u64) =
+                    (Height::from(h), MaxThreadCount::from(*t), u);
+
+                // tree build compute time
+                group.bench_function(
+                    BenchmarkId::new(
+                        "build_tree",
+                        format!("{:?}/{:?}/NUM_USERS: {:?}", &tup.0, &tup.1, &tup.2),
+                    ),
+                    |bench| {
+                        bench.iter(|| {
+                            e.advance().unwrap();
+                            ndm_smt = Some(setup::build_ndm_smt(tup.clone()));
+                        });
+                    },
+                );
+
+                // tree build memory usage
+                let alloc = alloc.read().unwrap();
+                let act = act.read().unwrap();
+                let res = res.read().unwrap();
+
+                let mem_usage = MemoryUsage {
+                    allocated: setup::bytes_as_string(alloc),
+                    active: setup::bytes_as_string(act),
+                    resident: setup::bytes_as_string(res),
+                };
+
+                // tree build file size
+                let tree_build_file_size = setup::serialize_tree(
+                    ndm_smt.as_ref().expect("Tree not found"),
+                    PathBuf::from("./target"),
+                );
+
+                let tree_build = Metrics {
+                    variable: Variable::TreeBuild,
+                    mem_usage,
+                    file_size: tree_build_file_size,
+                };
+
+                println!("\n{:?}\n", tree_build);
+
+                let alloc = stats::allocated::mib().unwrap();
+                let act = stats::active::mib().unwrap();
+                let res = stats::resident::mib().unwrap();
+
+                let mut proof = Option::<InclusionProof>::None;
+
+                let entity_keys = ndm_smt.as_ref().unwrap().entity_mapping().keys();
+                let mut entity_ids: Vec<&EntityId> = Vec::new();
+
+                e.advance().unwrap();
+
+                entity_keys.for_each(|entity| {
+                    e.advance().unwrap();
+                    entity_ids.push(entity);
+                });
+
+                // proof generation compute time
+                group.bench_function(
```

Comment:
This bench + the verification one use the same tree & entity_id so there is no variation in the 10 or so runs that criterion does. A better statistical test would use different entity_ids to do each run. You should be able to do this easily using `bench_with_input`. Best to make a different group for each of tree building, proof generation, and proof verification.